### PR TITLE
Handle deriving fmt::Debug impl for bitfields instead of std

### PR DESCRIPTION
**Pull request by jam1garner**
Implementation for my suggestion from #31.

### CODE REVIEW

**Comment by jam1garner** [comment]
Clippy absolutely calling me out. Ident derefs to `str`? Ugh I wish I knew that months ago.

File: impl/src/bitfield.rs [hunk 1]
```diff
@@ -271,22 +325,96 @@ impl BitfieldStruct {
...
+            #[allow(clippy::identity_op, unused_attributes)]
             #vis struct #ident
             {
                 bytes: [::core::primitive::u8; #next_divisible_by_8 / 8usize],
             }
         )
     }
 
+    /// Get whether or not `derive(Debug)` is included in the attributes
+    fn should_derive_debug(&self) -> bool {
```
Comment (Robbepop) [inline impl/src/bitfield.rs:345]: We should also provide a conditional `Debug` impl for cases like
```rust
#[bitfield]
#[cfg_attr(test, derive(Debug))]
pub struct MyBitfield { ... }
```
In this case we'd like to also only implement `Debug` if the given conditional (in this case `test`) is true.

File: impl/src/bitfield.rs [hunk 2]
```diff
@@ -271,22 +325,96 @@ impl BitfieldStruct {
...
+                        false
+                    }
+                }
+                _ => false,
+            }
+        })
+    }
+
+    /// Get the attributes with `Debug` filtered out of any present derives.
+    fn get_attrs_no_debug<'a>(&'a self) -> impl Iterator<Item = TokenStream2> + 'a {
```
Comment (Robbepop) [inline impl/src/bitfield.rs:376]: This should be extended to work with the mentioned `cfg_attr`.

File: impl/src/bitfield.rs [hunk 3]
```diff
@@ -180,6 +184,56 @@ impl BitfieldStruct {
...
+                                    #field_strings,
+                                    self.#field_getters()
+                                        .as_ref()
+                                        .map(|field| field as &dyn core::fmt::Debug)
+                                        .unwrap_or_else(|err| err as &dyn core::fmt::Debug)
+                                )
+                            )*
+                            .finish()
+                    }
+                }
```
Comment (Robbepop) [inline impl/src/bitfield.rs:230]:
```suggestion
                impl ::core::fmt::Debug for #ident {
                    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
                        f.debug_struct(::core::stringify!(#ident))
                            #(
                                .field(
                                    #field_strings,
                                    self.#field_getters()
                                        .as_ref()
                                        .map(|field| field as &dyn ::core::fmt::Debug)
                                        .unwrap_or_else(|err| err as &dyn ::core::fmt::Debug)
                                )
                            )*
                            .finish()
                    }
                }
```

Comment (Robbepop) [reply impl/src/bitfield.rs:230]: proc. macro hygiene suggestion

**Review by Robbepop — COMMENTED** [review]
Thanks a lot for the PR! So hot! :fire: 

It is missing some test cases in the form of our `trybuild` UI tests.
Also it misses conditional compilation support with `cfg_attr` derives.
Some hygiene problems but nothing too scary.

All in all looks decent.

File: impl/src/bitfield.rs [hunk 4]
```diff
@@ -180,6 +184,56 @@ impl BitfieldStruct {
...
+            .iter()
+            .enumerate()
+            .map(|(i, field)| {
+                let field_name = field
+                    .ident
+                    .as_ref()
+                    .map(syn::Ident::to_string)
+                    .unwrap_or_else(|| format!("get_{}", i));
+                let fallible_getter =
+                    quote::format_ident!("{}_or_err", field_name, span = field.span());
```
Comment (Robbepop) [inline impl/src/bitfield.rs:201]: This is not the correct getter identifier for tuple structs with unnamed fields.
For those it is: `get_N_or_err` where `N` denotes the index of the field.

File: impl/src/bitfield.rs [hunk 5]
```diff
@@ -271,22 +325,96 @@ impl BitfieldStruct {
...
+        let attrs = self.get_attrs_no_debug();
         let vis = &self.item_struct.vis;
         let ident = &self.item_struct.ident;
         let size = self.generate_bitfield_size();
         let next_divisible_by_8 = Self::next_divisible_by_8(&size);
         quote_spanned!(span=>
             #( #attrs )*
             #[repr(transparent)]
-            #[allow(clippy::identity_op)]
+            #[allow(clippy::identity_op, unused_attributes)]
```
Comment (Robbepop) [inline impl/src/bitfield.rs:336]: why is `unused_attributes` needed here? what code spawns which warnings?

File: impl/src/bitfield.rs [hunk 6]
```diff
@@ -180,6 +184,56 @@ impl BitfieldStruct {
...
+        self.item_struct
+            .fields
+            .iter()
+            .enumerate()
+            .map(|(i, field)| {
+                let field_name = field
+                    .ident
+                    .as_ref()
+                    .map(syn::Ident::to_string)
+                    .unwrap_or_else(|| format!("get_{}", i));
```
Comment (Robbepop) [inline impl/src/bitfield.rs:199]: In this case we want to use the field's actual name as string, so we can strip away the `get_` prefix for unnamed fields.

**Comment by Robbepop** [comment]
As a test case you could take my example from the issue:
```rust
use modular_bitfield::prelude::*;

#[derive(BitfieldSpecifier)]
#[bits = 3]
pub enum Weekday { Monday, Tuesday, Wednesday, Thursday, Friday, Saturday, Sunday }

#[bitfield(specifier = true)]
pub struct Date {
    #[bits = 3]
    weekday: Weekday,
    week: u8,
}

#[bitfield]
pub struct FromTo {
    #[bits = 11]
    from: Date,
    #[bits = 11]
    to: Date,
    distance: B10,
}
```
The only thing you need in order to make it a real test is to add a `fn main() { .. }` that asserts the expected debug output.

File: impl/src/bitfield.rs (hunk 4 above)
Comment (jam1garner) [reply impl/src/bitfield.rs:201]: Ah this was a mistake of me trying to get too fancy when refactoring some stuff. Thanks!

File: impl/src/bitfield.rs (hunk 5 above)
Comment (jam1garner) [reply impl/src/bitfield.rs:336]: It's needed for empty derives. So `#[derive(Debug)]` becomes `#[derive()]` (which raises the warning). If you have strong feelings about going out of my way to detect and remove this though, I can handle it that way.

Comment (Robbepop) [reply impl/src/bitfield.rs:336]: Would it be possible to simply check if no derives are left and therefore skip re-expanding the whole derive attribute again?

File: impl/src/bitfield.rs [hunk 7]
```diff
@@ -271,22 +325,96 @@ impl BitfieldStruct {
...
+                    if is_derive {
+                        meta_list.nested = meta_list
+                            .nested
+                            .into_iter()
+                            .filter(|nested| {
+                                match nested {
+                                    syn::NestedMeta::Meta(syn::Meta::Path(path)) => {
+                                        path.get_ident()
+                                            .map(|ident| *ident != "Debug")
+                                            .unwrap_or(true)
```
Comment (Robbepop) [inline impl/src/bitfield.rs:395]: Should use https://docs.rs/syn/0.15.44/syn/struct.Path.html#method.is_ident here instead.

File: impl/src/bitfield.rs [hunk 8]
```diff
@@ -271,22 +325,96 @@ impl BitfieldStruct {
...
+    /// Get the attributes with `Debug` filtered out of any present derives.
+    fn get_attrs_no_debug<'a>(&'a self) -> impl Iterator<Item = TokenStream2> + 'a {
+        self.item_struct.attrs.iter().map(|attr| {
+            match attr.parse_meta() {
+                Ok(syn::Meta::List(mut meta_list)) => {
+                    let is_derive = meta_list
+                        .path
+                        .get_ident()
+                        .map(|ident| *ident == "derive")
+                        .unwrap_or(false);
```
Comment (Robbepop) [inline impl/src/bitfield.rs:384]: Should use https://docs.rs/syn/0.15.44/syn/struct.Path.html#method.is_ident instead.

File: impl/src/bitfield.rs (hunk 1 above)
Comment (Robbepop) [reply impl/src/bitfield.rs:345]: While implementing this in my own I found out that the Rust compiler already handles this for us. So nothing needs to be done for `cfg_attr` anywhere fortunately. :)
See also: https://github.com/Robbepop/modular-bitfield/pull/38

**Comment by Robbepop** [comment]
For inspiration on how to do the checks and extractions for attributes on the `#[bitfield]` struct and re-expansions take a look at:
https://github.com/Robbepop/modular-bitfield/blob/robin-implement-repr/impl/src/bitfield.rs#L121

File: impl/src/bitfield.rs (hunk 7 above)
Comment (Robbepop) [reply impl/src/bitfield.rs:395]:
```suggestion
                                        path.is_ident("Debug")
```

File: impl/src/bitfield.rs (hunk 8 above)
Comment (Robbepop) [reply impl/src/bitfield.rs:384]:
```suggestion
                    let is_derive = meta_list.path.is_ident("derive");
```

**Comment by Robbepop** [comment]
I am closing this now since I pulled your commits over to some local branch in order to integrate your derive implementation to the new config architecture. Thanks for the PR and work! Your commits are going to be included in the commit log to honor your work. :tada:

**Comment by Robbepop** [comment]
I have created the follow-up PR here: https://github.com/Robbepop/modular-bitfield/pull/39 @jam1garner

**Pull request closed without merging**